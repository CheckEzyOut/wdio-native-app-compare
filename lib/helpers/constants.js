export const ACTUAL = 'actual';
export const DEFAULT_IMAGE_FORMAT_NAME = '{tag}-{deviceName}';
export const DEFAULT_BASELINE_FODLER = './baseline';
export const DEFAULT_SCREENSHOTS = './.tmp/screenshots';
export const DIFF = 'diff';
export const NO_DEVICE_NAME = 'no-device-name-in-caps';
export const STATUS_BAR_SELECTORS = {
    IOS: '//XCUIElementTypeStatusBar',
    ANDROID: '//*[@resource-id="android:id/statusBarBackground"]'
};
