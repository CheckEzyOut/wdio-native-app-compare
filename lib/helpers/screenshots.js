/**
 * Save a screenshot to a file
 *
 * @return {Promise<*>}
 */
export async function getBufferedScreenshot() {
    return new Buffer(await browser.saveScreenshot(), 'base64');
}

/**
 * Save a screenshot to a file
 *
 * @param {string} filePath
 *
 * @return {Promise<*>}
 */
export async function saveToFile(filePath) {
    return await browser.saveScreenshot(filePath);
}
