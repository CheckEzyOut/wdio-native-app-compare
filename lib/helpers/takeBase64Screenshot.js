/**
 * Take a screenshot
 *
 * @return {Promise<string>}
 */
export async function takeBase64Screenshot() {
    return await Buffer.from(await browser.saveScreenshot()).toString('base64');
}
