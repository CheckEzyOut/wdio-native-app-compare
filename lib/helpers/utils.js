import { ensureDirSync } from 'fs-extra';
import { resolve } from 'path';
import { NO_DEVICE_NAME } from './constants';

/**
 * Create the filename
 *
 * @param {string} formatString
 * @param {string} tag
 *
 * @return {string}
 */
export function formatFileName(formatString, tag) {
    const capabilities = browser.desiredCapabilities;
    const defaults = {
        deviceName: capabilities.deviceName || NO_DEVICE_NAME,
        orientation: capabilities.orientation || 'no-orientation-in-caps',
        platformName: capabilities.platformName || 'no-platform-name-in-caps',
        platformVersion: capabilities.platformVersion || 'no-platform-version-in-caps',
        tag,
    };

    Object.keys(defaults).forEach((value) => {
        formatString = formatString.replace(`{${value}}`, defaults[value]);
    });

    return formatString.replace(/ /g, '_').toLowerCase() + '.png';
}

/**
 * Determine the file path
 *
 * @param {string}  folder
 * @param {boolean} savePerDevice
 * @param {string}  fileName
 *
 * @return {string}
 */
export function determineFilePath (folder, savePerDevice, fileName){
    const deviceFolder = browser.desiredCapabilities.deviceName
        ? browser.desiredCapabilities.deviceName.replace(/ /g, '_').toLowerCase()
        : NO_DEVICE_NAME ;
    const folderName = resolve(process.cwd(), folder, savePerDevice ? deviceFolder : '');

    ensureDirSync(folderName);

    return resolve(folderName, fileName);
}

/**
 * Get the device info and store it to a global if needed
 *
 * @param   {buffer}  bufferedScreenshot
 *
 * @return  {Promise<{
 *      dpr: number,
 *      screensize: {
 *          height: number,
 *          width: number,
 *      },
 *      screenshotHeight: number,
 *      screenshotWidth: number
 * }>}
 */
export async function getDeviceInfo(bufferedScreenshot) {
    if (browser.wdioNac.hasOwnProperty('deviceInfo')) {
        return browser.wdioNac.deviceInfo;
    }

    const screenshotWidth = bufferedScreenshot.readUInt32BE(16);
    const screensize = (await browser.windowHandleSize()).value;
    const deviceInfo = {
        dpr: screenshotWidth / screensize.width,
        screensize,
        screenshotHeight: bufferedScreenshot.readUInt32BE(20),
        screenshotWidth,
    };

    browser.wdioNac.deviceInfo = deviceInfo;

    return deviceInfo;
}

/**
 * Return the values of an object multiplied against the devicePixelRatio
 *
 * @param   {object} values
 * @param   {number} dpr
 *
 * @return  {object}
 */
export function multiplyObjectValuesAgainstDPR(values, dpr) {
    Object.keys(values).map(value => {
        values[value] *= dpr;
    });

    return values;
}
