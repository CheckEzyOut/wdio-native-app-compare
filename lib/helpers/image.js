import {resolve} from 'path';
import { access, copySync, createWriteStream, readFileSync, writeFileSync } from 'fs-extra';

const PNGImage = require('png-image');

import { multiplyIosObjectValuesAgainstDPR } from './utils';

/**
 * Check if the given image already exists
 *
 * @param {object}  folders
 * @param {string}  fileName
 * @param {boolean} autoSaveBaseline
 *
 * @return {Promise<any>}
 */
export async function checkImageExists(folders, fileName, autoSaveBaseline) {
    return new Promise((resolvePromise, rejectPromise) => {
        access(resolve(folders.baseline, fileName), (error) => {
            if (error) {
                if (autoSaveBaseline) {
                    try {
                        copySync(
                            resolve(folders.actual, fileName),
                            resolve(folders.baseline, fileName),
                        );
                        console.log(`\nINFO: Autosaved the image to ${resolve(folders.baseline, fileName)}\n`);
                    } catch (err) {
                        rejectPromise(`Image could not be copied. The following error was thrown: ${err}`);
                    }
                } else {
                    rejectPromise('Image not found, saving current image as new baseline.');
                }
            }
            resolvePromise();
        });
    });
}

/**
 * Save a (cropped) screen
 *
 * @param {buffer}  bufferedScreenshot
 * @param {string}  fileNamePath
 * @param {object}  rectangles
 * @param {object}  deviceInfo
 *
 * @return {Promise}
 */
export async function saveCroppedScreenshot(bufferedScreenshot, fileNamePath, rectangles, deviceInfo) {
    const rect = multiplyIosObjectValuesAgainstDPR(rectangles, deviceInfo.dpr);
    return new PNGImage({
        imagePath: bufferedScreenshot,
        imageOutputPath: fileNamePath,
        cropImage: rect
    }).runWithPromise();
}

/**
 * Save the image
 *
 * @param {string}  filePath
 * @param {buffer}  image
 */
export function saveImage(filePath, image) {
    writeFileSync(filePath, image);
}
