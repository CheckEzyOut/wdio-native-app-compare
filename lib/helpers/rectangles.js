import { STATUS_BAR_SELECTORS } from './constants';
import { getDeviceInfo, multiplyIosObjectValuesAgainstDPR } from './utils';

/**
 * Determine the ignore rectangles
 *
 * @param {buffer}  bufferedScreenshot
 * @param {object}  options
 *
 * @return {Promise<Array>}
 */
export async function determineIgnoreRectangles(bufferedScreenshot, options) {
    let rectangles = [];
    const deviceInfo = await getDeviceInfo(bufferedScreenshot);

    if (options.blockOutStatusBar) {
        rectangles.push(await determineStatusBarRectangles(deviceInfo.dpr));
    }

    if (options.blockOuts.length > 0) {
        rectangles = rectangles.concat(determineBlockOuts(options.blockOuts, deviceInfo.dpr));
    }

    if (options.elementBlockOuts.length > 0) {
        rectangles = rectangles.concat(await determineElementBlockOuts(options.elementBlockOuts, deviceInfo.dpr));
    }

    return rectangles;
}

/**
 * Determine the status bar height for iOS/Android
 *
 * @param {number} dpr The device pixel ratio
 *
 * @return {Promise<{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }>}
 */
async function determineStatusBarRectangles(dpr) {
    let allowInvisibleElements;

    if (browser.wdioNac.hasOwnProperty('statusBarDimensions')) {
        return browser.wdioNac.statusBarDimensions;
    }

    // With Android and UiAutomator2 the status bar is not in the view by default
    // It can be seen when {settings: {allowInvisibleElements: true}} is set
    // See http://appium.io/docs/en/advanced-concepts/settings/
    if (browser.isAndroid) {
        allowInvisibleElements = (await browser.settings()).value.allowInvisibleElements;
        await browser.settings({ allowInvisibleElements: true });
    }

    const statusBarSize = await browser.getElementSize(STATUS_BAR_SELECTORS[browser.isIOS ? 'IOS' : 'ANDROID']);
    if (browser.isAndroid) {
        await browser.settings({ allowInvisibleElements });
    }
    const statusBarDimensions = multiplyIosObjectValuesAgainstDPR({
        height: statusBarSize.height,
        width: statusBarSize.width,
        x: 0,
        y: 0
    }, dpr);

    browser.wdioNac.statusBarDimensions = statusBarDimensions;

    return statusBarDimensions;
}

/**
 * Determine the block outs data
 *
 * @param {array}   blockOuts
 * @param {number}  dpr The device pixel ratio
 *
 * @return {Promise<[{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }]>}
 */
function determineBlockOuts(blockOuts, dpr) {
    const dimensions = [];
    const validKeys = ['x', 'y', 'width', 'height'];

    blockOuts.forEach(blockOut => {
        let validBlockOutObject = true;
        Object.keys(blockOut).forEach(key => {
            if (!validKeys.includes(key)) {
                console.log(`
###################################################
 Blockout:
 ${JSON.stringify(blockOut)}

 is not allowed. Key: ${key} is not allowed, 
 only 'x', 'y', 'width', 'height' are allowed
###################################################
`);
                validBlockOutObject = false;
            }
        });

        if (validBlockOutObject) {
            dimensions.push(blockOut);
        }
    });

    return dimensions.map(dimension => multiplyIosObjectValuesAgainstDPR(dimension, dpr));
}

/**
 * Determine the element block out data
 *
 * @param {array}   elements
 * @param {number}  dpr The device pixel ratio
 *
 * @return {Promise<[{
 *      height: number,
 *      width: number,
 *      x: number,
 *      y: number
 * }]>}
 */
async function determineElementBlockOuts(elements, dpr) {
    const dimensions = [];

    for (let element of elements) {
        const selector = await element.selector;
        const index = await element.index;
        const size = await browser.getElementSize(selector);
        const coordinates = await browser.getLocation(selector);

        dimensions.push({
            height: Array.isArray(size) ? size[index].height : size.height,
            width: Array.isArray(size) ? size[index].width : size.width,
            x: Array.isArray(coordinates) ? coordinates[index].x : coordinates.x,
            y: Array.isArray(coordinates) ? coordinates[index].y : coordinates.y,
        });
    }

    return dimensions.map(dimension => multiplyIosObjectValuesAgainstDPR(dimension, dpr));
}
