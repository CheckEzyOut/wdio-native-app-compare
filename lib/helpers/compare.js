import { readFileSync } from 'fs-extra';
import { join, resolve } from 'path';
import { compareImages } from '../resemble/compareImages';
import { saveImage } from './image';

/**
 * Compare 2 images with each other
 *
 * @param {object}  folders
 * @param {string}  fileName
 * @param {object}  options
 *
 * @return {Promise<{
 *      fileName: {string},
 *      folders: {object},
 *      misMatchPercentage: number
 * }>}
 */
export async function executeCompare(folders, fileName, options) {
    if (options.ignoreAlpha) {
        options.ignore.push('alpha')
    }
    if (options.ignoreAntialiasing) {
        options.ignore.push('antialiasing')
    }
    if (options.ignoreColors) {
        options.ignore.push('colors')
    }
    if (options.ignoreLess) {
        options.ignore.push('less')
    }
    if (options.ignoreNothing) {
        options.ignore.push('nothing')
    }

    // Execute the compare and retrieve the data
    const data = await compareImages(
        readFileSync(resolve(folders.baseline, fileName)),
        readFileSync(resolve(folders.actual, fileName)),
        options
    );
    const misMatchPercentage = options.rawMisMatchPercentage
        ? data.rawMisMatchPercentage
        : Number(data.rawMisMatchPercentage.toFixed(2));

    if (misMatchPercentage > options.saveAboveTolerance || options.debug) {
        saveImage(resolve(folders.diff, fileName), data.getBuffer());
    }

    return {
        fileName,
        folders,
        misMatchPercentage,
    };
}

/**
 * Determine the default compare options
 *
 * @param {object}  options
 * @param {boolean} options.blockOutStatusBar If the statusbar on mobile / tablet needs to blocked out by default
 * @param {boolean} options.blockOutNavigationBar If the navigation bar on mobile / tablet needs to blocked out by default. This will only work on Android
 * @param {boolean} options.elementBlockOuts One or multiple elements that need to be blocked outs
 * @param {boolean} options.debug Add some extra logging and always save the image difference (default:false)
 * @param {boolean} options.ignoreAlpha compare images and discard alpha
 * @param {boolean} options.ignoreAntialiasing compare images and discard anti aliasing
 * @param {boolean} options.ignoreColors Even though the images are in colour, the comparison wil compare 2 black/white images
 * @param {boolean} options.ignoreLess compare images and compare with red = 16, green = 16, blue = 16, alpha = 16, minBrightness=16, maxBrightness=240
 * @param {boolean} options.ignoreNothing compare images and compare with red = 0, green = 0, blue = 0, alpha = 0, minBrightness=0, maxBrightness=255
 * @param {boolean} options.ignoreTransparentPixel Will ignore all pixels that have some transparency in one of the images
 * @param {number}  options.largeImageThreshold Enable skipping pixels during the comparison to mitigate performance issues based on a amount of pixels (width / height), default 0
 * @param {boolean} options.rawMisMatchPercentage default false. If true the return percentage will be like 0.12345678, default is 0.12
 * @param {number}  options.saveAboveTolerance Allowable value of misMatchPercentage that prevents saving image with differences
 *
 * @return {{
 *      blockOutStatusBar: boolean,
 *      blockOutNavigationBar: boolean,
 *      debug: boolean,
 *      elementBlockOuts: array,
 *      ignore: array,
 *      ignoreAlpha: boolean,
 *      ignoreAntialiasing: boolean,
 *      ignoreColors: boolean,
 *      ignoreLess: boolean,
 *      ignoreNothing: boolean,
 *      ignoreRectangles: array,
 *      ignoreTransparentPixel: boolean,
 *      largeImageThreshold: number,
 *      rawMisMatchPercentage: boolean,
 *      saveAboveTolerance: number,
 *  }}
 */
export function defaultCompareOptions(options) {
    return {
        blockOuts: [],
        blockOutStatusBar: options.blockOutStatusBar || false,
        blockOutNavigationBar: options.blockOutNavigationBar || false,
        elementBlockOuts: options.elementBlockOuts || [],
        debug: options.debug || false,
        ignore: [],
        ignoreAlpha: options.ignoreAlpha || false,
        ignoreAntialiasing: options.ignoreAntialiasing || false,
        ignoreColors: options.ignoreColors || false,
        ignoreLess: options.ignoreLess || false,
        ignoreNothing: options.ignoreNothing || false,
        ignoreRectangles: [],
        ignoreTransparentPixel: options.ignoreTransparentPixel || false,
        largeImageThreshold: options.largeImageThreshold || 0,
        rawMisMatchPercentage: options.rawMisMatchPercentage || false,
        saveAboveTolerance: options.saveAboveTolerance || 0,
    }
}

/**
 * Determine the instance compare options
 *
 * @param {object}  options
 * @param {boolean} options.blockOuts One or multiple block outs on a screen / element can be provided in an Array with objects
 * @param {boolean} options.blockOutStatusBar If the statusbar on mobile / tablet needs to blocked out by default
 * @param {boolean} options.blockOutNavigationBar If the navigation bar on mobile / tablet needs to blocked out by default. This will only work for Android
 * @param {boolean} options.elementBlockOuts One or multiple elements that need to be blocked out
 * @param {boolean} options.ignoreAlpha compare images and discard alpha
 * @param {boolean} options.ignoreAntialiasing compare images and discard anti aliasing
 * @param {boolean} options.ignoreColors Even though the images are in colour, the comparison wil compare 2 black/white images
 * @param {boolean} options.ignoreLess compare images and compare with red = 16, green = 16, blue = 16, alpha = 16, minBrightness=16, maxBrightness=240
 * @param {boolean} options.ignoreNothing compare images and compare with red = 0, green = 0, blue = 0, alpha = 0, minBrightness=0, maxBrightness=255
 * @param {boolean} options.ignoreTransparentPixel Will ignore all pixels that have some transparency in one of the images
 *
 * @return {{}}
 */
export function instanceCompareOptions(options) {
    return {
        ...(options.blockOuts ? { blockOuts: options.blockOuts } : {}),
        ...(options.blockOutStatusBar ? { blockOutStatusBar: options.blockOutStatusBar } : {}),
        ...(options.blockOutNavigationBar ? { blockOutNavigationBar: options.blockOutNavigationBar } : {}),
        ...(options.elementBlockOuts ? { elementBlockOuts: options.elementBlockOuts } : {}),
        ...(options.ignoreAlpha ? { ignoreAlpha: options.ignoreAlpha } : {}),
        ...(options.ignoreAntialiasing ? { ignoreAntialiasing: options.ignoreAntialiasing } : {}),
        ...(options.ignoreColors ? { ignoreColors: options.ignoreColors } : {}),
        ...(options.ignoreLess ? { ignoreLess: options.ignoreLess } : {}),
        ...(options.ignoreNothing ? { ignoreNothing: options.ignoreNothing } : {}),
        ...(options.ignoreTransparentPixel ? { ignoreTransparentPixel: options.ignoreTransparentPixel } : {}),
    }
}
