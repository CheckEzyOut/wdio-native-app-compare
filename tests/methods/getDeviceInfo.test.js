import {IMAGE_STRING} from '../mocks/mocks'
// import {getDeviceInfo} from '../../lib/methods/getDeviceInfo'
import * as Utils from '../../lib/helpers/utils'

describe('getDeviceInfo', () => {
    let determineIphoneXSeriesSpy, determineLargeIphoneXSeriesSpy, getDeviceInfo, getScreenshotSizeSpy
    const screenData = {
        width: 10,
        height: 20,
    }

    beforeEach(() => {
        jest.isolateModules(() => {
            getDeviceInfo = require('../../lib/methods/getDeviceInfo').getDeviceInfo
        })
        delete global.browser
        global.browser = {
            getWindowSize: jest.fn().mockResolvedValue(screenData),
        }
        getScreenshotSizeSpy = jest.spyOn(Utils, 'getScreenshotSize').mockReturnValue({
            width: 1000,
            height: 2000,
        })
    })

    afterEach(() => {
        global.browser = {
            getWindowSize: jest.fn().mockRestore(),
        }
        determineIphoneXSeriesSpy.mockRestore()
        determineLargeIphoneXSeriesSpy.mockRestore()
        getScreenshotSizeSpy.mockRestore()
    })

    it('should be able to get the default device info for the initial run', async () => {
        determineIphoneXSeriesSpy = jest.spyOn(Utils, 'determineIphoneXSeries').mockReturnValue(false)
        determineLargeIphoneXSeriesSpy = jest.spyOn(Utils, 'determineLargeIphoneXSeries').mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).toBeCalledWith(IMAGE_STRING)
        expect(global.browser.getWindowSize).toHaveBeenCalled()
        expect(determineIphoneXSeriesSpy).toBeCalledWith(screenData)
        expect(determineLargeIphoneXSeriesSpy).toBeCalledWith(screenData)
    })

    it('should be able to get the default device info for the second run', async () => {
        determineIphoneXSeriesSpy = jest.spyOn(Utils, 'determineIphoneXSeries').mockReturnValue(false)
        determineLargeIphoneXSeriesSpy = jest.spyOn(Utils, 'determineLargeIphoneXSeries').mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).toBeCalledWith(IMAGE_STRING)
        expect(global.browser.getWindowSize).toHaveBeenCalled()
        expect(determineIphoneXSeriesSpy).toBeCalledWith(screenData)
        expect(determineLargeIphoneXSeriesSpy).toBeCalledWith(screenData)

        getScreenshotSizeSpy.mockRestore()
        determineIphoneXSeriesSpy.mockRestore()
        determineLargeIphoneXSeriesSpy.mockRestore()
        global.browser = {
            getWindowSize: jest.fn().mockRestore(),
        }

        // The second run to check that the data is stored in the `DEVICE_INFO` and all methods are not called again
        global.browser = {
            getWindowSize: jest.fn().mockResolvedValue(screenData),
        }
        getScreenshotSizeSpy = jest.spyOn(Utils, 'getScreenshotSize').mockReturnValue({
            width: 1000,
            height: 2000,
        })
        determineIphoneXSeriesSpy = jest.spyOn(Utils, 'determineIphoneXSeries').mockReturnValue(false)
        determineLargeIphoneXSeriesSpy = jest.spyOn(Utils, 'determineLargeIphoneXSeries').mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
        expect(getScreenshotSizeSpy).not.toHaveBeenCalled()
        expect(global.browser.getWindowSize).not.toHaveBeenCalled()
        expect(determineIphoneXSeriesSpy).not.toHaveBeenCalled()
        expect(determineLargeIphoneXSeriesSpy).not.toHaveBeenCalled()
    })

    it('should be able to get the device info for an iPhone X', async () => {
        determineIphoneXSeriesSpy = jest.spyOn(Utils, 'determineIphoneXSeries').mockReturnValue(true)
        determineLargeIphoneXSeriesSpy = jest.spyOn(Utils, 'determineLargeIphoneXSeries').mockReturnValue(false)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for an iPhone X Large', async () => {
        determineIphoneXSeriesSpy = jest.spyOn(Utils, 'determineIphoneXSeries').mockReturnValue(false)
        determineLargeIphoneXSeriesSpy = jest.spyOn(Utils, 'determineLargeIphoneXSeries').mockReturnValue(true)

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })

    it('should be able to get the device info for a landscape phone', async () => {
        global.browser = {
            getWindowSize: jest.fn().mockResolvedValue({
                width: 1000,
                height: 500,
            }),
        }

        expect(await getDeviceInfo(IMAGE_STRING)).toMatchSnapshot()
    })
})
