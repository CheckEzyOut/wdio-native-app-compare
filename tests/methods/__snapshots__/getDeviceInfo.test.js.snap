// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeviceInfo should be able to get the default device info for the initial run 1`] = `
Object {
  "dpr": 100,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
    },
    "STATUS_BAR": 20,
  },
  "isIphoneXSeries": false,
  "isLargeIphoneXSeries": false,
  "isPortrait": true,
  "screenSize": Object {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the default device info for the second run 1`] = `
Object {
  "dpr": 100,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
    },
    "STATUS_BAR": 20,
  },
  "isIphoneXSeries": false,
  "isLargeIphoneXSeries": false,
  "isPortrait": true,
  "screenSize": Object {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the default device info for the second run 2`] = `
Object {
  "dpr": 100,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
    },
    "STATUS_BAR": 20,
  },
  "isIphoneXSeries": false,
  "isLargeIphoneXSeries": false,
  "isPortrait": true,
  "screenSize": Object {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for a landscape phone 1`] = `
Object {
  "dpr": 1,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
    },
    "STATUS_BAR": 20,
  },
  "isIphoneXSeries": false,
  "isLargeIphoneXSeries": false,
  "isPortrait": false,
  "screenSize": Object {
    "height": 500,
    "width": 1000,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone X 1`] = `
Object {
  "dpr": 100,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "bottom": 6,
        "left": 120,
        "right": 135,
        "top": 799,
      },
    },
    "STATUS_BAR": 44,
  },
  "isIphoneXSeries": true,
  "isLargeIphoneXSeries": false,
  "isPortrait": true,
  "screenSize": Object {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;

exports[`getDeviceInfo should be able to get the device info for an iPhone X Large 1`] = `
Object {
  "dpr": 100,
  "iosRectangles": Object {
    "HOME_BAR": Object {
      "PORTRAIT": Object {
        "bottom": 6,
        "left": 133,
        "right": 148,
        "top": 883,
      },
    },
    "STATUS_BAR": 44,
  },
  "isIphoneXSeries": false,
  "isLargeIphoneXSeries": true,
  "isPortrait": true,
  "screenSize": Object {
    "height": 20,
    "width": 10,
  },
  "screenshotHeight": 2000,
  "screenshotWidth": 1000,
}
`;
